mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/notation/UCINotationCombinators.kt

```diff
@@ -0,0 +1,49 @@
+package ch.epfl.sdp.mobile.application.chess.notation
+
+import ch.epfl.sdp.mobile.application.chess.engine.Rank
+import ch.epfl.sdp.mobile.application.chess.engine.rules.Action
+import ch.epfl.sdp.mobile.application.chess.engine.rules.Action.Move
+import ch.epfl.sdp.mobile.application.chess.engine.rules.Action.Promote
+import ch.epfl.sdp.mobile.application.chess.notation.GenericNotationCombinators.position
+import ch.epfl.sdp.mobile.application.chess.notation.GenericNotationCombinators.spaces
+import ch.epfl.sdp.mobile.application.chess.parser.Combinators.combine
+import ch.epfl.sdp.mobile.application.chess.parser.Combinators.flatMap
+import ch.epfl.sdp.mobile.application.chess.parser.Combinators.map
+import ch.epfl.sdp.mobile.application.chess.parser.Combinators.repeat
+import ch.epfl.sdp.mobile.application.chess.parser.Parser
+import ch.epfl.sdp.mobile.application.chess.parser.StringCombinators
+import ch.epfl.sdp.mobile.application.chess.parser.StringCombinators.checkFinished
+
+/** An object which contains some convenience parser combinators for algebraic notation. */
+object UCINotationCombinators {
+
+  /** A [Parser] which returns the [Rank] of a piece for promotion */
+  private val promoteRank =
+      combine(
+          StringCombinators.char('q').map { Rank.Queen },
+          StringCombinators.char('r').map { Rank.Rook },
+          StringCombinators.char('b').map { Rank.Bishop },
+          StringCombinators.char('n').map { Rank.Knight },
+      )
+
+  /** A [Parser] for a [Move] action. */
+  private val move = position.flatMap { from -> position.map { to -> Move(from, to) } }
+
+  /** A [Parser] for a [Promote] action. */
+  private val promote =
+      position.flatMap { from ->
+        position.flatMap { to -> promoteRank.map { rank -> Promote(from, to, rank) } }
+      }
+
+  /** A [Parser] for an action. */
+  private val action = combine(promote, move)
+
+  /** A [Parser] for an arbitrarily long string of actions. */
+  private val actions =
+      action.flatMap { action -> spaces.map { action } }.repeat().flatMap { actions ->
+        action.map { actions + it }
+      }
```

Comment:
Question : should the parser handle lists with zero items ?

Reply:
Although there would be no reason to be this way, there is also no reason for it to not be. I'll address it.

Reply:
Addressed in [67a6ba8](https://github.com/epfl-SDP/android/pull/283/commits/67a6ba8e69ab70d1449265ec841c3786d91bddf5).